test/common/config/subscription_factory_test.cc

```diff
@@ -194,51 +302,66 @@ TEST_P(SubscriptionFactoryTestApiConfigSource, NonExistentCluster) {
+  api_config_source->add_cluster_names("foo");
   Upstream::ClusterManager::ClusterInfoMap cluster_map;
   Upstream::MockCluster cluster;
-  cluster_map.emplace("eds_cluster", cluster);
+  cluster_map.emplace("foo", cluster);
   EXPECT_CALL(cm_, clusters()).WillOnce(Return(cluster_map));
-  EXPECT_CALL(cluster, info());
-  EXPECT_CALL(*cluster.info_, addedViaApi()).WillOnce(Return(true));
-  EXPECT_THROW_WITH_MESSAGE(subscriptionFromConfigSource(config)->start({"foo"}, callbacks_),
-                            EnvoyException,
-                            "envoy::api::v2::core::ConfigSource must have a statically defined "
-                            "non-EDS cluster: 'eds_cluster' "
-                            "does not exist, was added via api, or is an EDS cluster");
+  if (api_config_source->api_type() == envoy::api::v2::core::ApiConfigSource::GRPC) {
+    EXPECT_THROW_WITH_MESSAGE(
+        subscriptionFromConfigSource(config)->start({"foo"}, callbacks_), EnvoyException,
+        "envoy::api::v2::core::ConfigSource::GRPC must not have a cluster name specified");
+  } else {
+    EXPECT_CALL(cluster, info());
+    EXPECT_CALL(*cluster.info_, addedViaApi()).WillOnce(Return(true));
+    EXPECT_THROW_WITH_MESSAGE(
+        subscriptionFromConfigSource(config)->start({"foo"}, callbacks_), EnvoyException,
+        "envoy::api::v2::core::ConfigSource must have a statically defined "
+        "non-EDS cluster: 'foo' does not exist, was added via api, or is an EDS cluster");
+  }
 }
 
 TEST_P(SubscriptionFactoryTestApiConfigSource, EDSClusterBackingEDSCluster) {
   envoy::api::v2::core::ConfigSource config;
   auto* api_config_source = config.mutable_api_config_source();
   api_config_source->set_api_type(GetParam());
-  api_config_source->add_cluster_names("eds_cluster");
+  api_config_source->add_cluster_names("foo");
   Upstream::ClusterManager::ClusterInfoMap cluster_map;
   Upstream::MockCluster cluster;
-  cluster_map.emplace("eds_cluster", cluster);
+  cluster_map.emplace("foo", cluster);
   EXPECT_CALL(cm_, clusters()).WillOnce(Return(cluster_map));
-  EXPECT_CALL(cluster, info()).Times(2);
-  EXPECT_CALL(*cluster.info_, addedViaApi());
-  EXPECT_CALL(*cluster.info_, type()).WillOnce(Return(envoy::api::v2::Cluster::EDS));
-  EXPECT_THROW_WITH_MESSAGE(subscriptionFromConfigSource(config)->start({"foo"}, callbacks_),
-                            EnvoyException,
-                            "envoy::api::v2::core::ConfigSource must have a statically defined "
-                            "non-EDS cluster: 'eds_cluster' "
-                            "does not exist, was added via api, or is an EDS cluster");
+  if (api_config_source->api_type() == envoy::api::v2::core::ApiConfigSource::GRPC) {
+    EXPECT_THROW_WITH_MESSAGE(
+        subscriptionFromConfigSource(config)->start({"foo"}, callbacks_), EnvoyException,
+        "envoy::api::v2::core::ConfigSource::GRPC must not have a cluster name specified");
```

Comment:
I think we want to setup this test to actually get the below behavior, by using grpc_services.envoy_grpc.cluster_name to specify cluster name in the gRPC case.

Reply:
I'll refactor all the `SubscriptionFactoryTestApiConfigSource/*` tests to better fit this.